In [1]:
# Numerical Packages
import numpy as np
import pandas as pd

#Common packages
import copy
from tqdm import tqdm
import re 
import glob
 
#Plot's Making  Packages
import matplotlib.pyplot as plt
from matplotlib.ticker import MaxNLocator
from mpl_toolkits.axes_grid1 import make_axes_locatable

#System Packages
import importlib
import time
import os
from os.path import isdir, isfile, join
import logging

importlib.reload(logging)
logging.basicConfig(level = logging.INFO)


# Learning packages
import sklearn
from sklearn.utils import shuffle

logging.info("Numpy Version: {}".format(np.__version__))
logging.info("Pnadas Version: {}".format(pd.__version__))
logging.info("Sklearn Version: {}".format(sklearn.__version__))

%config InlineBackend. figure_format = 'retina'

INFO:root:Numpy Version: 1.19.5
INFO:root:Pnadas Version: 1.1.5
INFO:root:Sklearn Version: 0.24.1


# Combine High-Level-Feature Data from CSV file

In [2]:
%%time 
def High_Level_Features(csv_file=[]):
    
    if len(csv_file) < 1:
        raise Inputerror("Please check high-level features files!!")
        
    
    high_level_feature = pd.read_csv(csv_file[0])
    high_level_feature = high_level_feature[(high_level_feature["PTJ_0"] > 300 ) & (high_level_feature["PTJ_0"] <= 480 )]
#     high_level_feature = high_level_feature[(high_level_feature["PTJ_0"] > 300 )]
    logging.info(str(csv_file[0]))
    logging.info("File Length: {}".format(len(high_level_feature)))
    logging.info("Higgs jet Pt min: {:.3f},  Higgs jet Pt max: {:.3f}".format(high_level_feature["PTJ_0"].min(),high_level_feature["PTJ_0"].max()))
    logging.info("\r")
    
    for i, file in enumerate(csv_file):
        if i == 0:
            continue
        else:
            logging.info(str(file))
            
            if "250_500" in str(file):

                dataframe = pd.read_csv(file)
                dataframe = dataframe[(dataframe["PTJ_0"] > 300) & (dataframe["PTJ_0"] <= 480)]
            
            elif "450_700" in str(file):

                dataframe = pd.read_csv(file)
                dataframe = dataframe[(dataframe["PTJ_0"] > 480) & (dataframe["PTJ_0"] <= 700)]
                
            elif "650_900" in str(file):

                dataframe = pd.read_csv(file)
                dataframe = dataframe[(dataframe["PTJ_0"] > 700) & (dataframe["PTJ_0"] <= 900)]
            
            elif "850_1100" in str(file):

                dataframe = pd.read_csv(file)
                dataframe = dataframe[(dataframe["PTJ_0"] > 900) & (dataframe["PTJ_0"] <= 1100)]
                
                
                
                
#             dataframe = dataframe[(dataframe["PTJ_0"] > 300 )]
            high_level_feature = pd.concat([high_level_feature, dataframe], ignore_index=True, axis=0,join='inner')
            high_level_feature["WEIGHT"] = high_level_feature["weight"]/(len(csv_file)//4)
        
        
        logging.info("File Length: {}".format(len(dataframe)))
        logging.info("Higgs jet Pt min: {:.3f},  Higgs jet Pt max: {:.3f}".format(dataframe["PTJ_0"].min(),dataframe["PTJ_0"].max()))
        logging.info("\r")
    
    logging.info( "\033[3;43m Total File Length: {} \033[0;m".format(len(high_level_feature)))
    logging.info("\r")
    
    return high_level_feature


"""
Local and Data Path @ASGC
"""
path = "/dicos_ui_home/alanchung/Universality_Boosetd_Higgs/Data_ML/Data_High_Level_Features"
    
    
process = {
            "herwig_ang_H" : 0,
            "herwig_ang_QCD" : 0,
            "pythia_def_H" : 0,
            "pythia_def_QCD" : 0,
            "pythia_vin_H" : 0,
            "pythia_vin_QCD" : 0,
            "pythia_dip_H" : 0,
            "pythia_dip_QCD" : 0 ,
#                 "sherpa_def_H" : 0,
#                 "sherpa_def_QCD" : 0
              }  

try:
    
    for index, element in tqdm(enumerate(process)):
        
#         if index%2 == 0:
            
        process[element] = pd.read_csv(path + "/" + str(element) + ".csv")

        print("{}'s event number: {}".format(element,len(process[element])))
            
#         if index%2 == 1:
            
#             process[element] = pd.read_csv(path + "/" + str(element) + ".csv")
            
#             process[element] = pd.concat([process[element] for i in range(10)],ignore_index=True,axis=0,join='inner')
#             process[element].drop(columns='index')
#             process[element]["index"]= np.linspace(0,len(process[element])-1,len(process[element]),dtype="int")
#             process[element].to_csv(path + "/" + str(element) + ".csv",index = 0)
            
#             print("{}'s event number: {}".format(element,len(process[element])))
        
        
        
except:

    for index, element in tqdm(enumerate(process)):
        print("Combing Data for {}".format(element))

        csv_files = sorted(glob.glob(path+"/"+str(element)+"_*.csv"))
        process[element] = High_Level_Features(csv_files)
        process[element]["index"]= np.linspace(0,len(process[element])-1,len(process[element]),dtype="int")
        process[element].to_csv(path + "/" + str(element) + ".csv",index = 0)

        
        
        if index%2 == 1:
            
            process[element] = pd.read_csv(path + "/" + str(element) + ".csv")
            
            process[element] = pd.concat([process[element] for i in range(20)],ignore_index=True,axis=0,join='inner')
            process[element].drop(columns='index')
            process[element]["index"]= np.linspace(0,len(process[element])-1,len(process[element]),dtype="int")
            process[element].to_csv(path + "/" + str(element) + ".csv",index = 0)
            
            
            print("{}'s event number: {}".format(element,len(process[element])))
        
        
    
feature = process["pythia_def_H"].columns
print("Features: {}".format(feature))



1it [00:02,  2.55s/it]

herwig_ang_H's event number: 907695


2it [00:10,  4.05s/it]

herwig_ang_QCD's event number: 2056380


3it [00:13,  3.82s/it]

pythia_def_H's event number: 908409


4it [00:22,  5.47s/it]

pythia_def_QCD's event number: 2834720


5it [00:25,  4.65s/it]

pythia_vin_H's event number: 868948


6it [00:31,  5.20s/it]

pythia_vin_QCD's event number: 2098580


7it [00:34,  4.35s/it]

pythia_dip_H's event number: 876286


8it [00:43,  5.44s/it]

pythia_dip_QCD's event number: 3160420
Features: Index(['GEN', 'SHO', 'PRO', 'MJ_0', 'PTJ_0', 'eta_0', 'phi_0', 't21_0',
       'D21_0', 'D22_0', 'C21_0', 'C22_0', 'MJ', 'PTJ', 'eta', 'phi', 't21',
       'D21', 'D22', 'C21', 'C22', 'weight', 'eventindex', 'WEIGHT', 'index'],
      dtype='object')
CPU times: user 39.7 s, sys: 3.77 s, total: 43.5 s
Wall time: 43.5 s


 12%|█▎        | 1/8 [00:02<00:16,  2.43s/it]

herwig_ang_H's event number: 907695

 25%|██▌       | 2/8 [00:02<00:11,  1.84s/it]

herwig_ang_QCD's event number: 102819

 38%|███▊      | 3/8 [00:05<00:10,  2.09s/it]

pythia_def_H's event number: 908409

 50%|█████     | 4/8 [00:06<00:06,  1.67s/it]

pythia_def_QCD's event number: 141736

 62%|██████▎   | 5/8 [00:08<00:05,  1.88s/it]

pythia_vin_H's event number: 868948

 75%|███████▌  | 6/8 [00:09<00:02,  1.48s/it]

pythia_vin_QCD's event number: 104929

 88%|████████▊ | 7/8 [00:11<00:01,  1.79s/it]

pythia_dip_H's event number: 876286

100%|██████████| 8/8 [00:12<00:00,  1.57s/it]

pythia_dip_QCD's event number: 158021

# Generate Image Lists 

In [ ]:
HOMEPATH = "/dicos_ui_home/alanchung/UniversalityDNN_DiHiggs/"
JetImagePath =  HOMEPATH + "Data_ML/" +"Image_Directory_Leading/"
savepath = HOMEPATH + "Data_ML/"

process = {
            "herwig_ang_H" : 0,
            "herwig_ang_QCD" : 0,
            "pythia_def_H" : 0,
            "pythia_def_QCD" : 0,
            "pythia_vin_H" : 0,
            "pythia_vin_QCD" : 0,
            "pythia_dip_H" : 0,
            "pythia_dip_QCD" : 0 ,
#             "sherpa_def_H" : 0,
#             "sherpa_def_QCD" : 0
              }  

    
try:
    
    for element in tqdm(process):
        total_data = pd.read_csv(savepath + str(element) + "_dict.csv")
    
    
        print("List path: {}".format(savepath + str(element) + "_dict.csv"))
        print("{}'s event number: {}".format(element,len(total_data)))
        

except:
    
    try: 
        folder_list_tmp = os.listdir(JetImagePath)

        for label, element in tqdm(enumerate(process)):
            img_length = 0
            path_jet_image = []
            folder_list = []

            for i, folder in enumerate(folder_list_tmp):
                    if isdir( join(JetImagePath, folder)):
                       
                        if folder[:12] == str(element):
                            folder_list.append([folder,int(folder[17:])])


                        if folder[:14] == str(element):
                            folder_list.append([folder,int(folder[19:])])

            folder_list = sorted(folder_list, key = lambda s: s[1])

            for i, folder in enumerate(folder_list):
                
                if isdir( join(JetImagePath, folder[0])):

                    if label%2 == 0:
                        if folder[0][:12] == str(element):
                            images_folder_tmp = os.listdir(join(JetImagePath, folder[0]))

                            images_folder = sorted(images_folder_tmp, key = lambda s: int(s[6:len(s)-4]))

                            for image in images_folder:
                                path_jet_image.append(folder[0] + "/" + image)

                            img_length += len(images_folder)

                    if label%2 == 1:
                        if folder[0][:14] == str(element):
                            images_folder_tmp = os.listdir(join(JetImagePath, folder[0]))
                            images_folder = sorted(images_folder_tmp, key = lambda s: int(s[6:len(s)-4]))

                            for image in images_folder:
                                path_jet_image.append(folder[0] + "/" + image)

                            img_length += len(images_folder)
                         
                    
            if label%2 == 0:
                Ydata = np.full(img_length, 1) # Higgs Jet
            if label%2 == 1:
                Ydata = np.full(img_length, 0) # QCD Jet

            dataframe = pd.DataFrame()
            dataframe["JetImage"] = path_jet_image
            dataframe["Y"] = np.array(Ydata,dtype="int")
            dataframe.to_csv(savepath + str(element) + "_dict.csv",index = 0)
            
            print("List path: {}".format(savepath  + str(element) + "_dict.csv"))
            print("{}'s event number: {}".format(element,len(dataframe)))
    except:
        
        print("error")
    

# Split Data into Training, Validation and Test dataset


In [12]:
HOMEPATH = "/dicos_ui_home/alanchung/Universality_Boosetd_Higgs/"
Data_High_Level_Features_path =  HOMEPATH + "Data_ML/" +"Data_High_Level_Features/"
 
total_list = ['GEN', 'SHO', 'PRO', 'MJ_0', 'PTJ_0', 'eta_0', 'phi_0', 't21_0',
       'D21_0', 'D22_0', 'C21_0', 'C22_0', 'MJ', 'PTJ', 'eta', 'phi', 't21',
       'D21', 'D22', 'C21', 'C22', 'weight', 'eventindex', 'WEIGHT', 'index']

# if preprocess = "trimmed":

#     features =  ["MJ","PTJ","t12","D21","D22","C21","C22"]
    
# elif if preprocess = "untrimmed":
    
#      features =  ["MJ_0","PTJ_0","t12_0","D21_0","D22_0","C21_0","C22_0"]

if os.path.exists(HOMEPATH + "Data_ML/BDT") == 0:
    os.mkdir(HOMEPATH + "Data_ML/BDT/")
    datapath = HOMEPATH + "Data_ML/Data_High_Level_Features/"
#     savepath = HOMEPATH + "Data_ML/BDT/"
    savepath = HOMEPATH +  "Notebook/KFold/"
    
else: 
    datapath = HOMEPATH + "Data_ML/Data_High_Level_Features/"
#     savepath = HOMEPATH + "Data_ML/BDT/"
    savepath = HOMEPATH +  "Notebook/KFold/"



#######################################################################################
print(time.strftime("%a %b %d %H:%M:%S %Y", time.localtime()))
ticks_1 = time.time()    




try:
    
    data_train = {
#             "herwig_ang_train" : 0,
            "pythia_def_train" : 0,
#             "pythia_vin_train" : 0,
#             "pythia_dip_train" : 0,
#             "sherpa_def_train" : 0
            }  
    
    data_test = {
#             "herwig_ang_test" : 0,
            "pythia_def_test" : 0,
#             "pythia_vin_test" : 0,
#             "pythia_dip_test" : 0,
#             "sherpa_def_test" : 0
            }  
    
    data_val = {
#             "herwig_ang_val" : 0,
            "pythia_def_val" : 0,
#             "pythia_vin_val" : 0,
#             "pythia_dip_val" : 0,
#             "sherpa_def_val" : 0
            }  
    
    for i, element in enumerate(data_train):
        data_train[element] = pd.read_csv(savepath + str(element) + ".csv")
        
    for i, element in enumerate(data_test):
        data_test[element] = pd.read_csv(savepath + str(element) + ".csv")
        
    for i, element in enumerate(data_val):
        data_val[element] = pd.read_csv(savepath + str(element) + ".csv")

    print("All Files are loaded!")

    ticks_2 = time.time()
    totaltime =  ticks_2 - ticks_1
    print("\033[3;33mTime Cost : {:.4f} min\033[0;m".format(totaltime/60.))
    #######################################################################################

    print("H jet : QCD jet = 1 : 1")
    print("\r")
#     print("{:^8}{:^15}".format("","Pythia Default"))
#     print("{:^8}{:^15}".format("Train #",len(data_train["pythia_def_train"])))
#     print("{:^8}{:^15}".format("Test #",len(data_test["pythia_def_test"])))
#     print("{:^8}{:^15}".format("Val. #",len(data_val["pythia_def_val"])))

    train = [ len(data_train[element]) for j, element in enumerate(data_train)]
    test = [len(data_test[element]) for j, element in enumerate(data_test)]
    val = [len(data_val[element]) for j, element in enumerate(data_val)]
    
    print("{:^8}{:^15}{:^15}{:^15}".format("","Herwig Angular","Pythia Vincia","Pythia Dipole"))
    print("{:^8}{:^15}{:^15}{:^15}".format("Train #",train[0],train[1],train[2]))
    print("{:^8}{:^15}{:^15}{:^15}".format("Test #",test[0],test[1],test[2]))
    print("{:^8}{:^15}{:^15}{:^15}".format("Val. #",val[0],val[1],val[2]))
    
    

except:

    process = {
            "herwig_ang_H" : 0,
            "herwig_ang_QCD" : 0,
            "pythia_def_H" : 0,
            "pythia_def_QCD" : 0,
            "pythia_vin_H" : 0,
            "pythia_vin_QCD" : 0,
            "pythia_dip_H" : 0,
            "pythia_dip_QCD" : 0 ,
#                 "sherpa_def_H" : 0,
#                 "sherpa_def_QCD" : 0
              }  
    
    
    for i, element in enumerate(process):
        process[element] = pd.read_csv(datapath + str(element) + ".csv")
        

    for j, element in enumerate(process):
        print(element, len(process[element]))
    

    """
    Mass Cut and PT cut
    """
    #######################
    for j , element in enumerate(process):
        tmp = process[element]
        
#         """
#         Leading Jet Cuts
#         """
#         tmp = tmp[(tmp["MJ1_0"] > 50) ]
#         tmp = tmp[(tmp["PTJ1_0"] > 450)]
#         tmp = tmp[(abs(tmp["eta1_0"]) < 2)]
        
#         """
#         Leading Jet Cuts with Higgs Window
#         """
#         tmp = tmp[(tmp["MJ_0"] > 110) ]
#         tmp = tmp[(tmp["MJ_0"] < 160) ]
        

#         """
#         SubLeading Jet Cuts with Higgs Window
#         """
#         tmp = tmp[(tmp["MJ2_0"] > 110) ]
#         tmp = tmp[(tmp["MJ2_0"] < 160) ]
        
#         """
#         Full Cuts
#         """
#         tmp = tmp[(tmp["MJ1_0"] > 50) & (tmp["MJ2_0"] > 50)]
#         tmp = tmp[(tmp["PTJ1_0"] > 450) & (tmp["PTJ2_0"] > 250)]
#         tmp = tmp[(abs(tmp["eta1_0"]) < 2) & (abs(tmp["eta2_0"]) < 2)]
#         tmp = tmp[(abs(tmp["delta_eta_0"]) < 1.3)]

        
        """
        Combine Leading and Subleading Jet
        """
        tmp = tmp[(tmp["PTJ_0"] > 300) ]
        tmp = tmp[(tmp["PTJ_0"] < 500) ]
        tmp = tmp[(tmp["MJ_0"] > 110) ]
        tmp = tmp[(tmp["MJ_0"] < 160) ]

        
        print("{} Preselection Efficiency: {:.3f}".format(element,len(tmp)/len(process[element])))
        process[element] = tmp
    #######################
    
    print("All Files are loaded!")
    
    H_length = [ len(process[element]) for j, element in enumerate(process) if j%2 == 0]
    QCD_length = [len(process[element]) for j, element in enumerate(process) if j%2 == 1]
#     print(H_length)
#     print(QCD_length)
#     print("{:^8}{:^15}{:^15}{:^15}{:^15}{:^15}".format("","Herwig Angular","Pythia Default","Pythia Vincia","Pythia Dipole","Sherpa Default"))
#     print("{:^8}{:^15}{:^15}{:^15}{:^15}{:^15}".format("H #",H_length[0],H_length[1],H_length[2],H_length[3],H_length[4]))
#     print("{:^8}{:^15}{:^15}{:^15}{:^15}{:^15}".format("QCD #",len(herwig_ang_QCD),len(pythia_def_QCD),len(pythia_vin_QCD),len(pythia_dip_QCD),len(sherpa_def_QCD)))
    
    print("{:^8}{:^15}{:^15}{:^15}{:^15}".format("","Herwig Angular","Pythia Default","Pythia Vincia","Pythia Dipole"))
    print("{:^8}{:^15}{:^15}{:^15}{:^15}".format("H #",H_length[0],H_length[1],H_length[2],H_length[3]))
    print("{:^8}{:^15}{:^15}{:^15}{:^15}".format("QCD #",QCD_length[0],QCD_length[1],QCD_length[2],QCD_length[3]))

#     print("{:^8}{:^15}{:^15}{:^15}".format("","Herwig Angular","Pythia Vincia","Pythia Dipole"))
#     print("{:^8}{:^15}{:^15}{:^15}".format("H #",H_length[0],H_length[1],H_length[2]))
#     print("{:^8}{:^15}{:^15}{:^15}".format("QCD #",QCD_length[0],QCD_length[1],QCD_length[2]))


 

    ticks_2 = time.time()
    totaltime =  ticks_2 - ticks_1
    print("\033[3;33mTime Cost : {:.4f} min\033[0;m".format(totaltime/60.))
    print("\r")
    

    for j , element in enumerate(process):
        process[element] = shuffle(process[element])

    print("All Files are shuffled!")
    ticks_2 = time.time()
    totaltime =  ticks_2 - ticks_1
    print("\033[3;33mTime Cost : {:.4f} min\033[0;m".format(totaltime/60.))
    print("\r")
    

    print("Prepare Training Files")
    ticks_2 = time.time()
    totaltime =  ticks_2 - ticks_1
    print("\033[3;33mTime Cost : {:.4f} min\033[0;m".format(totaltime/60.))
    print("\r")
    
    length = np.min([ len(process[element]) for element in process])
    
    data_train = {
            "herwig_ang_train" : 0,
            "pythia_def_train" : 0,
            "pythia_vin_train" : 0,
            "pythia_dip_train" : 0,
#             "sherpa_def_train" : 0
            }  
    
    data_test = {
            "herwig_ang_test" : 0,
            "pythia_def_test" : 0,
            "pythia_vin_test" : 0,
            "pythia_dip_test" : 0,
#             "sherpa_def_test" : 0
            }  
    
    data_val = {
            "herwig_ang_val" : 0,
            "pythia_def_val" : 0,
            "pythia_vin_val" : 0,
            "pythia_dip_val" : 0,
#             "sherpa_def_val" : 0
            }  

    
    ntrain = int(length*1)
    for i, element in enumerate(data_train):
        data_train[element] = pd.concat([ process[tmp][:ntrain] for j, tmp in enumerate(process) if (j >= i*2 and 2*i+1 >= j) ],ignore_index=True,axis=0,join='inner')

    ntest = int(length*0.) + ntrain
    for i, element in enumerate(data_test):
        data_test[element] = pd.concat([ process[tmp][ntrain:ntest] for j, tmp in enumerate(process) if (j >= i*2 and 2*i+1 >= j) ],ignore_index=True,axis=0,join='inner')
    
    nval = int(length*0.) + ntest
    for i, element in enumerate(data_val):
        data_val[element] = pd.concat([ process[tmp][ntest:nval] for j, tmp in enumerate(process) if (j >= i*2 and 2*i+1 >= j) ],ignore_index=True,axis=0,join='inner')


    print("Prepare Target Labels")
    ticks_2 = time.time()
    totaltime =  ticks_2 - ticks_1
    print("\033[3;33mTime Cost : {:.4f} min\033[0;m".format(totaltime/60.))
    print("\r")
    
    y_train = np.concatenate((np.full(int(length*1), 1), np.full(int(length*1), 0)))
    y_test = np.concatenate((np.full(int(length*0.), 1), np.full(int(length*0.), 0)))
    y_val = np.concatenate((np.full(int(length*0.), 1), np.full(int(length*0.), 0)))
    
    
    for i, element in enumerate(data_train):
        data_train[element]["target"] = y_train
    
    for i, element in enumerate(data_test):
        data_test[element]["target"] = y_test
        
    for i, element in enumerate(data_val):
        data_val[element]["target"] = y_val
        
        
    print("Save training, test and validation datasets")
    ticks_2 = time.time()
    totaltime =  ticks_2 - ticks_1
    print("\033[3;33mTime Cost : {:.4f} min\033[0;m".format(totaltime/60.))
    print("\r")
    
    for i, element in enumerate(data_train):
        data_train[element].to_csv(savepath + str(element) + ".csv",index = 0)
        
    for i, element in enumerate(data_test):
        data_test[element].to_csv(savepath + str(element) + ".csv",index = 0)
        
    for i, element in enumerate(data_val):
        data_val[element].to_csv(savepath + str(element) + ".csv",index = 0)

    print("All Files are saved!")
    ticks_2 = time.time()
    totaltime =  ticks_2 - ticks_1
    print("\033[3;33mTime Cost : {:.4f} min\033[0;m".format(totaltime/60.))
    print("\r")
    

    print("H jet : QCD jet = 1 : 1")
    print("\r")
    print("{:^8}{:^15}".format("","Pythia Default"))
    print("{:^8}{:^15}".format("Train #",len(data_train["pythia_def_train"])))
    print("{:^8}{:^15}".format("Test #",len(data_test["pythia_def_test"])))
    print("{:^8}{:^15}".format("Val. #",len(data_val["pythia_def_val"])))

    train = [ len(data_train[element]) for j, element in enumerate(data_train)]
    test = [len(data_test[element]) for j, element in enumerate(data_test)]
    val = [len(data_val[element]) for j, element in enumerate(data_val)]
    
    print("{:^8}{:^15}{:^15}{:^15}{:^15}".format("","Herwig Angular","Pythia Default","Pythia Vincia","Pythia Dipole"))
    print("{:^8}{:^15}{:^15}{:^15}{:^15}".format("Train #",train[0],train[1],train[2],train[3]))
    print("{:^8}{:^15}{:^15}{:^15}{:^15}".format("Test #",test[0],test[1],test[2],test[3]))
    print("{:^8}{:^15}{:^15}{:^15}{:^15}".format("Val. #",val[0],val[1],val[2],val[3]))

#     print("{:^8}{:^15}{:^15}{:^15}".format("","Herwig Angular","Pythia Vincia","Pythia Dipole"))
#     print("{:^8}{:^15}{:^15}{:^15}".format("Train #",train[0],train[1],train[2]))
#     print("{:^8}{:^15}{:^15}{:^15}".format("Test #",test[0],test[1],test[2]))
#     print("{:^8}{:^15}{:^15}{:^15}".format("Val. #",val[0],val[1],val[2]))


Fri Sep 10 02:57:00 2021
herwig_ang_H 907695
herwig_ang_QCD 2056380
pythia_def_H 908409
pythia_def_QCD 2834720
pythia_vin_H 868948
pythia_vin_QCD 2098580
pythia_dip_H 876286
pythia_dip_QCD 3160420
herwig_ang_H Preselection Efficiency: 0.201
herwig_ang_QCD Preselection Efficiency: 0.082
pythia_def_H Preselection Efficiency: 0.199
pythia_def_QCD Preselection Efficiency: 0.064
pythia_vin_H Preselection Efficiency: 0.197
pythia_vin_QCD Preselection Efficiency: 0.064
pythia_dip_H Preselection Efficiency: 0.199
pythia_dip_QCD Preselection Efficiency: 0.068
All Files are loaded!
        Herwig Angular Pythia Default  Pythia Vincia  Pythia Dipole 
  H #       182453         181068         171446         174530     
 QCD #      168740         181040         134380         215640     
Time Cost : 0.7053 min

All Files are shuffled!
Time Cost : 0.7096 min

Prepare Training Files
Time Cost : 0.7096 min

Prepare Target Labels
Time Cost : 0.7128 min

Save training, test and validation datasets
Time 

# Calculate Average Value and Variance for Image's Normalization and Standardization

In [6]:
# Take in a tuple of image lists, calculate average and variance
def Get_average_var(image_lists):
    tmp_av, tmp_var = np.zeros((3,40,40)), np.zeros((3,40,40))
    for i in range(3):
        tmp_av[i] = np.average(image_lists[:,i], axis=0)

        tmp_var[i] = np.var(image_lists[:,i], axis=0)

    return tmp_av, tmp_var


In [7]:
HOMEPATH = "/dicos_ui_home/alanchung/UniversalityDNN_DiHiggs/"
JetImagePath =  HOMEPATH + "Data_ML/" +"Image_Directory_Leading/"
savepath = HOMEPATH + "Data_ML/"

process = {
#             "herwig_ang_H" : 0,
#             "herwig_ang_QCD" : 0,
#             "pythia_def_H" : 0,
#             "pythia_def_QCD" : 0,
                "pythia_vin_H" : 0,
                "pythia_vin_QCD" : 0,
                "pythia_dip_H" : 0,
                "pythia_dip_QCD" : 0 ,
#                 "sherpa_def_H" : 0,
#                 "sherpa_def_QCD" : 0
              }  
    
for i, element in enumerate(process):
    process[element] = pd.read_csv(savepath + str(element) + "_dict.csv")


    
try:
    
    for i, element in enumerate(process):
        av = np.load(savepath + "average" + "_" + str(element) + ".npy")
        var = np.load(savepath + "variance" + "_" + str(element) + ".npy")
        print(str(element)+" average shape: ",av.shape)
        print(str(element)+" varianve shape: ",var.shape)
    
except:

    batch_size = 1000

    nan_path_list = []
    
    for i in range(len(process)):
        nan_path_list.append([])

    for i, element in enumerate(process):
        av, var = np.zeros((3,40,40)), np.zeros((3,40,40))
        length = 0
        nb_samples = len(process[element])
        for start in tqdm(range(0, nb_samples, batch_size)):

            x_batch = []
            end = min(start + batch_size, nb_samples)

            for img_path in range(start, end):

                x_jet_path = JetImagePath + process[element]["JetImage"].iloc[img_path]
                x_jet = np.load(x_jet_path)["jet_image"]

                if np.isnan(x_jet).any() == True:
    #             if np.nansum(x_train_jet) == 0:
                    nan_path_list[i].append(x_jet_path)
                    continue 
                x_batch.append(x_jet)

            x_batch = np.array(x_batch)

            tmp_av, tmp_var = Get_average_var(x_batch)
            av += tmp_av*len(x_batch)
            var += tmp_var
            length += len(x_batch)


    #         if start == 3000:
    #             break


        av = av/length
        np.save(savepath + "average" + "_" + str(element) + ".npy", av)
        np.save(savepath + "variance" + "_" + str(element) + ".npy", var)



pythia_vin_H average shape:  (3, 40, 40)
pythia_vin_H varianve shape:  (3, 40, 40)
pythia_vin_QCD average shape:  (3, 40, 40)
pythia_vin_QCD varianve shape:  (3, 40, 40)
pythia_dip_H average shape:  (3, 40, 40)
pythia_dip_H varianve shape:  (3, 40, 40)
pythia_dip_QCD average shape:  (3, 40, 40)
pythia_dip_QCD varianve shape:  (3, 40, 40)
